cpp/spglib.hpp

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include <array>
+#include <exception>
+#include <memory>
+#include <string>
+#include <tuple>
+#include <vector>
+
+namespace Spglib {
+	/**
+	 * Get the version of the Spglib C++ binding
+	 *
+	 * @return X.Y.Z version
+	 */
+	const std::string version();
+	/**
+	 * Get the full version of the Spglib C++ binding
+	 *
+	 * @return full version
+	 */
+	const std::string version_full();
+	/**
+	 * Get the commit of the Spglib C++ binding
+	 *
+	 * @return commit
+	 */
+	const std::string commit();
+	/**
+	 * C++ exception object for spglib related errors
+	 */
+	class SpglibError : public std::exception {
+		const std::string errorMsg;
+
+	public:
+		const char* what() const noexcept override;
+		SpglibError(std::string&& message);
+	};
+	class Symmetry {
+		// TODO: Find better interface compatible with std::mdspan
+		std::array<double, 9> rotations;
+		std::array<double, 3> translations;
+	};
+	class MagneticSymmetry : public Symmetry {
+		bool time_reversal;
+	};
+	class Atom {
+		std::array<double, 3> position;
+		int type;
+	};
+	// TODO: Better name?
+	class Lattice {
+		std::array<double, 9> lattice;
+		std::vector<Atom> atoms;
+		// TODO: Allow for non-periodic dimensions
+	};
+	class PrimitiveLattice : public Lattice {
+		Lattice& original_lattice;
+	};
+	class Spacegroup {
+		int number;
+		int hall_number;
+		std::string international_symbol;
+		std::string hall_symbol;
+		std::string schoenflies;
+		std::string pointgroup_international;
+		std::string pointgroup_schoenflies;
+		int arithmetic_crystal_class_number;
+		std::string arithmetic_crystal_class_symbol;
+		// TODO: Unclear what this is referencing
+		char choice[6];
+	};
```

Comment:
Needs review if these are sufficient or overcomplete.

Reply:
Space group and space-group-type have different meanings in crystallography. So here `SpacegroupType` is more appropriate.

Reply:
Good to know, what about the contents of it? E.g. `hall_number` and `hall_symbol` should be linked right?

Reply:
Do you ask about one-to-one correspondance between `hall_number` and `hall_symbol`?

Reply:
Yes, among other things. What makes sense to group together, what is a different convention of another, etc. The info would still be accessible flat, e.g. `soacegroup->hall_number()`, but it should be stored in more organized containers.

Reply:
Hall number is not a proper word in crystallography, but we in spglib use it as the serial ID of the Hall symbol.